### PR TITLE
Instance segmentation examples

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

Add examples that demonstrate how to train MaskFormer and Mask2Former for instance segmentation with Trainer or Accelerate.

Fixes #30335 
(processing void segmentation maps, I can make a separate PR, but for now its included)

At the moment kind of blocked by:
 - #30799 (not that critical, could be merged as is, as a workaround, I pass both values `reduce_labels` and `do_reduce_labels`)
 - #30943 (critical for instance segmentation, because masks are heavy and quickly occupy all RAM)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@amyeroberts @NielsRogge

### CODE REVIEW

File: examples/pytorch/instance-segmentation/README.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+```
+
+and boom, you're training, possibly on multiple GPUs, logging everything to all trackers found in your environment (like Weights and Biases, Tensorboard) and regularly pushing your model to the hub (with the repo name being equal to `args.output_dir` at your HF username) 🤗
+
+With the default settings, the script fine-tunes a [Mask2Former](https://huggingface.co/docs/transformers/model_doc/mask2former) model on the sample of [ADE20K](https://huggingface.co/datasets/qubvel-hf/ade20k-mini) dataset. The resulting model can be seen here: https://huggingface.co/qubvel-hf/finetune-instance-segmentation-ade20k-mini-mask2former-no-trainer. 
+
+
+## Reload and perform inference
+
+This means that after training, you can easily load your trained model and perform inference as follows::
```
Comment (qubvel) [inline examples/pytorch/instance-segmentation/README.md:129]:
```suggestion
This means that after training, you can easily load your trained model and perform inference as follows:
```

File: examples/pytorch/instance-segmentation/run_instance_segmentation.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+        args, training_args = parser.parse_args_into_dataclasses()
+
+    # Set default training arguments for instance segmentation
+    training_args.eval_do_concat_batches = False
+    training_args.batch_eval_metrics = True
+    training_args.remove_unused_columns = False
+
+    # # Sending telemetry. Tracking the example usage helps us better allocate resources to maintain them. The
+    # # information sent is the one passed as arguments along with your Python/PyTorch versions.
+    # send_example_telemetry("run_instance_segmentation", args)
```
Comment (qubvel) [inline examples/pytorch/instance-segmentation/run_instance_segmentation.py:368]:
```suggestion
    send_example_telemetry("run_instance_segmentation", args)
```

File: examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+
+    return repo_id
+
+
+def main():
+    args = parse_args()
+
+    # Sending telemetry. Tracking the example usage helps us better allocate resources to maintain them. The
+    # information sent is the one passed as arguments along with your Python/PyTorch versions.
+    # send_example_telemetry("run_instance_segmentation_no_trainer", args)
```
Comment (qubvel) [inline examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py:418]:
```suggestion
    send_example_telemetry("run_instance_segmentation_no_trainer", args)
```

File: examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+        ignore_mismatched_sizes=True,
+        token=args.hub_token,
+    )
+
+    image_processor = image_processor_class.from_pretrained(
+        args.model_name_or_path,
+        do_resize=True,
+        size={"height": args.image_height, "width": args.image_width},
+        do_reduce_labels=args.do_reduce_labels,
+        reduce_labels=args.do_reduce_labels,  # TODO: remove in the future
```
Comment (qubvel) [inline examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py:481]:
```suggestion
        reduce_labels=args.do_reduce_labels,  # TODO: remove when mask2former support `do_reduce_labels`
```

File: examples/pytorch/instance-segmentation/run_instance_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+        ignore_mismatched_sizes=True,
+        token=args.token,
+    )
+
+    image_processor = image_processor_class.from_pretrained(
+        args.model_name_or_path,
+        do_resize=True,
+        size={"height": args.image_height, "width": args.image_width},
+        do_reduce_labels=args.do_reduce_labels,
+        reduce_labels=args.do_reduce_labels,  # TODO: remove in the future
```
Comment (qubvel) [inline examples/pytorch/instance-segmentation/run_instance_segmentation.py:417]:
```suggestion
        reduce_labels=args.do_reduce_labels,  # TODO: remove when mask2former support `do_reduce_labels`
```

File: examples/pytorch/instance-segmentation/run_instance_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+        target_sizes = []
+        for target in post_processed_targets:
+            target_sizes.append(target["masks"].shape[-2:])
+        return target_sizes
+
+    def postprocess_prediction_batch(self, prediction_batch, target_sizes) -> List[Dict[str, torch.Tensor]]:
+        """Collect predictions in a form of list of dictionaries with keys "masks", "labels", "scores"."""
+
+        model_output = ModelOutput(class_queries_logits=prediction_batch[0], masks_queries_logits=prediction_batch[1])
+        post_processed_output = self.image_processor.post_process_instance_segmentation(
```
[the remark this answers is not in the document]
Comment (qubvel) [reply examples/pytorch/instance-segmentation/run_instance_segmentation.py:234]: Hi @SangbumChoi , thanks a lot for taking time to look into this  🤗 
Not sure I got your question, could you provide more details?

[the remark this answers is not in the document]
Comment (SangbumChoi) [reply examples/pytorch/instance-segmentation/run_instance_segmentation.py:234]: when the image size is very large such as 2000x3000, I faced this kind of process (such as resizing into original target size) took almost 10s for single image.

I think I need to test this script with large file :)

[the remark this answers is not in the document]
Comment (qubvel) [reply examples/pytorch/instance-segmentation/run_instance_segmentation.py:234]: Ok, got it. I observed that evaluation is relatively slow. However, we have to apply postprocessing in order to get metrics. In case it is too slow we can probably set a rare evaluation strategy (specify a large number of steps between evaluations) or completely drop evaluation during the training and rely on loss values only.
Another thing is to test different approaches to speed up this step, e.g. use lower resolution predicted masks, compute on GPU, or find more optimized operations for postprocessing.
Any help and feedback is appreciated 🤗

[the remark this answers is not in the document]
Comment (SangbumChoi) [reply examples/pytorch/instance-segmentation/run_instance_segmentation.py:234]: Yeah, that is also what I thought. I just wanted to make sure if there is any methods in find more optimized operations for postprocessing. Thanks for sharing :)

**Review by NielsRogge — APPROVED** [review]
This is great, thanks for adding this!

Feel free to also add a link to the example scripts in the "resources" section of the respective models

File: examples/pytorch/instance-segmentation/README.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+plt.show()
+```
+![Result](https://i.imgur.com/rZmaRjD.png)
+
+
+## Note on custom data
+
+Here is a short script demonstrating how to create your own dataset for instance segmentation and push it to the hub:
+
+> Note: Annotations should be represented as 3-channel images (similar to [scene_parsing_150](https://huggingface.co/datasets/zhoubolei/scene_parse_150#instance_segmentation-1) dataset), first channel is semantic-segmentation map with values corresponding to `label2id`, second is instance-segmentation map, where each instance has its unique value, third channel should be empty (filled with zeros).
```
[the remark this answers is not in the document]
Comment (qubvel) [reply examples/pytorch/instance-segmentation/README.md:199]: That's a fair point, I feel that your suggested way is more common.
But then they have to provide a class-id/name for each binary mask.

I think the best way would be to support various formats which provide different annotation platforms. On the other hand, usually platforms allow to export in different formats. Probably we can select several most common (like coco) and support them across segmentation, detection, instance segmentation.

**Comment by SangbumChoi** [comment]
@qubvel Hi pavel, One more question. How about adding DETR in this script since this is also returns instance output?
https://github.com/huggingface/transformers/blob/573565e35a5cc68f6cfb6337f5a93753ab16c65b/src/transformers/models/detr/image_processing_detr.py#L1774

File: examples/pytorch/instance-segmentation/run_instance_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+
+    # Setup logging and log on each process the small summary:
+    setup_logging(training_args)
+    logger.warning(
+        f"Process rank: {training_args.local_rank}, device: {training_args.device}, n_gpu: {training_args.n_gpu}, "
+        + f"distributed training: {training_args.parallel_mode.value == 'distributed'}, 16-bits training: {training_args.fp16}"
+    )
+    logger.info(f"Training/evaluation parameters {training_args}")
+
+    # Load last checkpoint form output_dir if it exists (and we are not overwriting it)
```
Comment (SangbumChoi) [inline examples/pytorch/instance-segmentation/run_instance_segmentation.py:378]:
```suggestion
    # Load last checkpoint from output_dir if it exists (and we are not overwriting it)
```

**Comment by qubvel** [comment]
> @qubvel Hi pavel, One more question. How about adding DETR in this script since this also returns instance output?

Hi @SangbumChoi it would be nice to have DETR also, but it's not straightforward since DETR's image processor has a different input annotation format and DETR ModelOutput has different fields it will require adding some conversion logic and I guess the example will be overcomplicated. I tried to keep it simple, but still not completely happy with the amount of code written and how much one should know to be able to modify it and adapt it for a particular use-case.

**Review by amyeroberts — APPROVED** [review]
Very nice! Thanks for working on and adding this!

**Comment by SangbumChoi** [comment]
> it would be nice to have DETR also, but it's not straightforward since DETR's image processor has a different input annotation format and DETR ModelOutput has different fields. It will require adding some conversion logic and I guess the example will be overcomplicated. I tried to keep it simple, but still not completely happy with the amount of code written and how much one should know to be able to modify it and adapt it for a particular use-case.

Exactly, I also try to make the logic simple as much as possible but unfortunately DETR has different annotation input :(
In the future, We can make seperate script for DETR only and just get link in here.

File: examples/pytorch/instance-segmentation/README.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+```
+
+and boom, you're training, possibly on multiple GPUs, logging everything to all trackers found in your environment (like Weights and Biases, Tensorboard) and regularly pushing your model to the hub (with the repo name being equal to `args.output_dir` at your HF username) 🤗
+
+With the default settings, the script fine-tunes a [Mask2Former](https://huggingface.co/docs/transformers/model_doc/mask2former) model on the sample of [ADE20K](https://huggingface.co/datasets/qubvel-hf/ade20k-mini) dataset. The resulting model can be seen here: https://huggingface.co/qubvel-hf/finetune-instance-segmentation-ade20k-mini-mask2former-no-trainer. 
+
+
+## Reload and perform inference
+
+This means that after training, you can easily load your trained model and perform inference as follows:
```
[the remark this answers is not in the document]
Comment (qubvel) [reply examples/pytorch/instance-segmentation/README.md]: Ok, thanks!

File: examples/pytorch/instance-segmentation/run_instance_segmentation.py [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+supported_models = {
+    "mask2former": {
+        "model": Mask2FormerForUniversalSegmentation,
+        "image_processor": Mask2FormerImageProcessor,
+    },
+    "maskformer": {
+        "model": MaskFormerForInstanceSegmentation,
+        "image_processor": MaskFormerImageProcessor,
+    },
+}
```
[the remark this answers is not in the document]
Comment (qubvel) [reply examples/pytorch/instance-segmentation/run_instance_segmentation.py]: Got it, fixed 👍

**Comment by qubvel** [comment]
Ok, I addressed most of the comments, thank you all for the reviews 🤗 

I am going to merge this:
 - At the moment, the problem with memory leaks is not critical because it occurs only with persistent DataLoader workers, which I believe is not a common case.
 - Support for other architectures (DETR, Conditional DETR) and dataset formats can be added as a follow-up.
